File: code/scenes/src/main/java/org/fbme/scenes/controllers/scene/SceneEditor.java
package org.fbme.scenes.controllers.scene;

import jetbrains.mps.openapi.editor.EditorContext;
import org.fbme.scenes.controllers.SceneViewpoint;
import org.jetbrains.annotations.NotNull;

import java.awt.event.MouseWheelListener;

public interface SceneEditor {

    EditorContext getContext();

    SceneViewpoint getImplicitViewpoint();

    void fireRelayout();

    void fireRepaint();

    void addCursorListener(@NotNull CursorListner listener);

    void removeCursorListener(@NotNull CursorListner listener);

    void addMouseWheelListener(@NotNull Layer layer, @NotNull MouseWheelListener listener);

    void removeMouseWheelListener(@NotNull MouseWheelListener listener);

    void addLayouter(@NotNull SceneLayouter layouter);

    void removeLayouter(@NotNull SceneLayouter layouter);

    void addCompletionProvider(@NotNull SceneCompletionProvider provider);

    void removeCompletionProvider(@NotNull SceneCompletionProvider provider);

    void addCellActionProvider(@NotNull SceneCellActionProvider provider);

    void removeCellActionProvider(@NotNull SceneCellActionProvider provider);

    @NotNull
    Layer createLayer(float apriority);

    void removeLayer(@NotNull Layer layer);

    void addClickListener(@NotNull Layer layer, @NotNull ClickEventListener listener);

    void removeClickListener(@NotNull ClickEventListener listener);

    void addDragListener(@NotNull Layer layer, @NotNull DragEventListener listener);

    void removeDragListener(@NotNull DragEventListener listener);

    void addCellProvider(@NotNull Layer layer, @NotNull EditorCellProvider provider);

    void removeCellProvider(@NotNull EditorCellProvider provider);

    void addPainter(@NotNull Layer layer, @NotNull ScenePainter painter);

    void removePainter(@NotNull ScenePainter painter);

    void addInitializer(@NotNull SceneInitializer initializer);

    void removeInitializer(@NotNull SceneInitializer initializer);

    <T> void storeState(SceneStateKey<T> key, T state);

    <T> T loadState(SceneStateKey<T> key);
}

File: code/scenes/src/main/kotlin/org/fbme/scenes/controllers/scene/SceneEditor.kt
package org.fbme.scenes.controllers.scene

import jetbrains.mps.openapi.editor.EditorContext
import org.fbme.scenes.controllers.SceneViewpoint
import java.awt.event.MouseWheelListener

interface SceneEditor {
    val editorContext: EditorContext
    val viewpoint: SceneViewpoint?
    fun fireRelayout()
    fun fireRepaint()
    fun addCursorListener(listener: CursorListener)
    fun removeCursorListener(listener: CursorListener)
    fun addMouseWheelListener(layer: Layer, listener: MouseWheelListener)
    fun removeMouseWheelListener(listener: MouseWheelListener)
    fun addLayouter(layouter: SceneLayouter)
    fun removeLayouter(layouter: SceneLayouter)
    fun addCompletionProvider(provider: SceneCompletionProvider)
    fun removeCompletionProvider(provider: SceneCompletionProvider)
    fun addCellActionProvider(provider: SceneCellActionProvider)
    fun removeCellActionProvider(provider: SceneCellActionProvider)
    fun createLayer(priority: Float): Layer
    fun removeLayer(layer: Layer)
    fun addClickListener(layer: Layer, listener: ClickEventListener)
    fun removeClickListener(listener: ClickEventListener)
    fun addDragListener(layer: Layer, listener: DragEventListener)
    fun removeDragListener(listener: DragEventListener)
    fun addCellProvider(layer: Layer, provider: EditorCellProvider)
    fun removeCellProvider(provider: EditorCellProvider)
    fun addPainter(layer: Layer, painter: ScenePainter)
    fun removePainter(painter: ScenePainter)
    fun addInitializer(initializer: SceneInitializer)
    fun removeInitializer(initializer: SceneInitializer)
    fun <T> storeState(key: SceneStateKey<T>, state: T)
    fun <T> loadState(key: SceneStateKey<T>): T?
}
